content/blog/prototyping/index.md

```diff
@@ -0,0 +1,732 @@
+There is support for dependencies as well.
+
+At the moment, `cargo-script` is a nightly feature, but it will be released soon on stable Rust.
+You can read more about it in the [RFC](https://rust-lang.github.io/rfcs/3424-cargo-script.html).
+
+### Don't Worry About Performance
+
+You have to try really really hard to write slow code in Rust.
+Use that to your advantage: during the prototype phase, try to keep the code as simple as possible; you can always optimize later.
+
+I gave a talk titled ["The Four Horsemen of Bad Rust Code"](https://github.com/corrode/four-horsemen-talk) where I
+argue that premature optimization is one of the biggest sins in Rust. 
+
+Especially experienced developers coming from C or C++ might be tempted to optimize too early.
+
+Rust makes code perform well by default - you get memory safety at virtually zero runtime cost. When developers try to optimize too early, they often fight the borrow checker by using complex lifetime annotations and intricate reference patterns in pursuit of better performance.
+This leads to harder-to-maintain code that may not actually run faster.
+
+Resist the urge to optimize too early.
+You will thank yourself later. [^1]
+
+[^1]: In the talk, I show an example where early over-optimization led to the wrong abstraction and made the code slower. The actual bottleneck was elsewhere and hard to uncover without profiling.
+
+### Use `println!` and `dbg!` for debugging 
+
+I find that printing values is pretty handy while prototyping.
+It's one less context switch to make compared to starting a debugger.
+
+Most people use `println!` for that, but [`dbg!`](https://doc.rust-lang.org/std/macro.dbg.html) has a few advantages:
+
+- It prints the file name and line number where the macro is called. This helps you quickly find the source of the output.
+- It outputs the expression as well as its value.
+- It's less syntax-heavy than `println!`; e.g. `dbg!(x)` vs. `println!("{x:?}")`.
+- It's only active in debug builds, so you don't have to worry about removing it later.
+
+Where `dbg!` really shines is in recursive functions or when you want to see the intermediate values during an iteration:
+
+```rust
+fn factorial(n: u32) -> u32 {
+    // `dbg!` returns the argument, 
+    // so you can use it in the middle of an expression
+    if dbg!(n <= 1) {
+        dbg!(1)
+    } else {
+        dbg!(n * factorial(n - 1))
+    }
+}
+
+dbg!(factorial(4));
+```
+
+The output is nice and tidy:
+
+```rust
+[src/main.rs:2:8] n <= 1 = false
+[src/main.rs:2:8] n <= 1 = false
+[src/main.rs:2:8] n <= 1 = false
+[src/main.rs:2:8] n <= 1 = true
+[src/main.rs:3:9] 1 = 1
+[src/main.rs:7:9] n * factorial(n - 1) = 2
+[src/main.rs:7:9] n * factorial(n - 1) = 6
+[src/main.rs:7:9] n * factorial(n - 1) = 24
+[src/main.rs:9:1] factorial(4) = 24
+```
+
+If you're interested, here are [more details on how to use that handy `dbg!` macro](https://edgarluque.com/blog/rust-dbg-macro/).
+
+### Play With Types
```

Comment:
I see the biggest advantage of the type system during prototyping that although it takes a bit longer to fidget with it, you end up having to do less iterations on the same part. Not sure I am making a clear point, so let's try: rust takes 3 units of time to iterate over how to structure data, python takes 2. However, with rust you need 2 iterations for 6 total units, with python you need 4 for a total of 8

Reply:
Wonder how I could phrase this. I'll give it a shot, but feel free to make a suggestion.

Reply:
I've integrated that into the text.